### ISSUE TITLE
ERROR TypeError: "item.toJSON is not a function

### ISSUE THREAD

**Issue by samantha99** [RicoSuter/NSwag#2862]
I have a parent object with a collection of children object:
```
public class EnduserCartModel
{
     ...
     public ICollection<ProductListModel> products {get;set;}
}
```

In controller :
```
[HttpPost("")]
... ([FromBody]EnduserModel enduserModel)
{
     ...
}
```

At the moment of sending the object from Angular to the Server I got this error: `ERROR TypeError: "item.toJSON is not a function.` This error appears on this line: `data["products"].push(item.toJSON());`

The code is generated by NSwag Studio: 13.6.0.0
.
So here the bug.

**Comment by karandesai28** [comment]
We are also facing this issue suddenly.

**Comment by thewax** [comment]
Am having the same issue.  ToJson with a collection of a collection. Seems to be working on an older build. I'm not exactly sure what's changed. If I find out, I'll reply with what I did.

**Comment by RicoSuter** [comment]
Seems to be a problem in NJsonSchema... can you create a repro with a unit test there?

**Comment by andmattia** [comment]
Hi have the same error (V. 13.5.0) but I found the problem in my case is related to a cast.

```typescript
args.columns = <EntityDto[]>columns; // Fail on toJSON
```
I see in debug that my object is not create from EntityDto and toJSON in not available
I change with this 
```typescript
 let colTyped : EntityDto[] = [];
        if(columns != undefined){
            for(var idx=0; idx < columns.length;idx++){
                var addCol = new EntityDto();
                addCol.technicalName = columns[idx].technicalName;
                colTyped.push(addCol);
            }
        }
```

And toJSON is valid function

**Comment by awdorrin** [comment]
I am encountering this issue in a project of mine with a nested class.
Like andmattia above, the only work around I have found, is to explicitly define each item in the array, to the subtype, before passing it to the generated client methods.

Looking through the generated code file, from what I can tell, the parent class that contains the nested class, is not provided the type of the nested object.

```
  toJSON(data?: any) {
        data = typeof data === 'object' ? data : {};
        if (Array.isArray(this.bids)) {
            data["bids"] = [];
            for (let item of this.bids)
                data["bids"].push(item.toJSON());
        }
```

so at the line: data["bids"].push(item.toJSON());
item is not of type 'BidDTO' in my case, so the code cannot resolve to the 'toJSON' function in the BidDTO class.

I am wondering, if the toJSON code that is generated could be made to do the 'new' during the push?
Perhaps something like:

```
            data["bids"] : BidDTO[]= [];
            for (let item of this.bids) {
                let temp = new BidDTO();
                Object.assign(temp, item);
                data["bids"].push(temp.toJSON());
            }        
```

**Comment by thewax** [comment]
@awdorrin I'm replying to you, but hopefully anyone else who sees this thread it can help them too: 

'BidDTO' needs to be an object. 

So when you .push(bidDTO) into an array, make sure you're not pushing a regular JSON object. 

var data = {};  //json

let bidDTO: BidDTO = data;

vs

let bidDTO: BidDTO = new BidDTO(data);

Only then will .ToJSON() be accessible. 

What was happening was:
- Array of SomeDTO comes from API.
- User creates new SomeDTO, which is added to list. However that SomeDTO was actually just a plain JSON and not instantiated as a class. 
- User saves, and we get the ToJSON is not a function error. 

So to fix that, always need to instantiate. We handle it this way in all our services that require adding new objects to an array.

**Comment by awdorrin** [comment]
Yeah, I get that, but I would think the NSwag generator should have enough information about the classes to create that code that instantiate the objects in the array itself.
It seems to handle the other direction properly.


On February 4, 2021 6:44:57 PM EST, thewax <notifications@github.com> wrote:
>@awdorrin I'm replying to you, but hopefully anyone else who sees this
>thread it can help them too: 
>
>'BidDTO' needs to be an object. 
>
>So when you .push(bidDTO) into an array, make sure you're not pushing a
>regular JSON object. 
>
>var data = {};  //json
>
>let bidDTO: BidDTO = data;
>
>vs
>
>let bidDTO: BidDTO = new BidDTO(data);
>
>Only then will .ToJSON() be accessible. 
>
>What was happening was:
>- Array of SomeDTO comes from API.
>- User creates new SomeDTO, which is added to list. However that
>SomeDTO was actually just a plain JSON and not instantiated as a class.
>
>- User saves, and we get the ToJSON is not a function error. 
>
>So to fix that, always need to instantiate. We handle it this way in
>all our services that require adding new objects to an array. 
>
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/RicoSuter/NSwag/issues/2862#issuecomment-773677463

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by slamj1** [comment]
@awdorrin, if you want NSwag to generate the correct code for the array type, perform the following steps:

1. Copy the liquid template file named ConvertToJavaScript.liquid from the C# project NJsonSchema.CodeGeneration.TypeScript from the GitHub repository RicoSuter/NJsonSchema into a directory of your choosing in your local project.
2. Modify ConvertToJavaScript.liquid, and replace (line 8):
 
`{{ Variable }}.push(item.toJSON());`

with:

`var obj{{ ArrayItemType }} = new {{ ArrayItemType }}(item);`
`{{ Variable }}.push(obj{{ ArrayItemType }}.toJSON());`

3. Modify your service.config.nswag file and tell the generator where your custom templates reside. For example:

`"templateDirectory": "./CustomTemplates",`

Now you should be good to go, and the item.toJSON() call will no longer throw an error.
The above template code also handles multiple array members.

@RicoSuter, could be a possible solution for the mainline template.

Cheers.

**Comment by slamj1** [comment]
@awdorrin, slight correction, make sure to have an open and closing brace around those two statements. Full template block (line 6 -10):

`for (let item of {{ Value }})`
`{%     if IsArrayItemNewableObject -%} {`
             `var obj{{ ArrayItemType }} = new {{ ArrayItemType }}(item);`
            `{{ Variable }}.push(obj{{ ArrayItemType }}.toJSON());`
    `}`

**Comment by lucaritossa** [comment]
This is a problem I'm facing today, in my project, after changed NSwag (v13.7.4) configuration from `"typeStyle": "Interface"` to `"Class"`

The problem is clear, all must be a defined object. 
@slamj1 proposed a solution that would be great if it was considered/evaluated as the mainline template.

But this issue seems to be staled.
Any news?

IMHO create a custom template could be a temporary work-around. I don't like this way because is not "upgrade-proof".

**Comment by Radwa-Moustafa** [comment]
I came across this issue multiple times and the solution was to make sure that your "item" is an instance of "ProductListModel" in front-end while you're appending this object to your array before sending it to the back end so when it calls .toJSON() in service-proxies.ts  it will find .toJSON() in your object and it will not throw that exception anymore.

**Comment by slamj1** [comment]
@lucaritossa, no news on this AFAIK. 

Agreed, a custom template is not exactly ideal as a workaround fro this paticular issue since the liquid template syntax does evolve.
